src/main/java/com/google/firebase/messaging/FirebaseMessaging.java

```diff
@@ -0,0 +1,399 @@
+  static final Map<Integer, String> IID_ERROR_CODES =
+      ImmutableMap.<Integer, String>builder()
+        .put(400, "invalid-argument")
+        .put(401, "authentication-error")
+        .put(403, "authentication-error")
+        .put(500, INTERNAL_ERROR)
+        .put(503, "server-unavailable")
+        .build();
+
+  private static final String IID_HOST = "https://iid.googleapis.com";
+  private static final String IID_SUBSCRIBE_PATH = "iid/v1:batchAdd";
+  private static final String IID_UNSUBSCRIBE_PATH = "iid/v1:batchRemove";
+
+  private final FirebaseApp app;
+  private final HttpRequestFactory requestFactory;
+  private final JsonFactory jsonFactory;
+  private final String url;
+
+  private HttpResponseInterceptor interceptor;
+
+  private FirebaseMessaging(FirebaseApp app) {
+    HttpTransport httpTransport = app.getOptions().getHttpTransport();
+    GoogleCredentials credentials = ImplFirebaseTrampolines.getCredentials(app);
+    this.app = app;
+    this.requestFactory = httpTransport.createRequestFactory(
+        new HttpCredentialsAdapter(credentials));
+    this.jsonFactory = app.getOptions().getJsonFactory();
+    String projectId = ImplFirebaseTrampolines.getProjectId(app);
+    checkArgument(!Strings.isNullOrEmpty(projectId),
+        "Project ID is required to access messaging service. Use a service account credential or "
+            + "set the project ID explicitly via FirebaseOptions. Alternatively you can also "
+            + "set the project ID via the GCLOUD_PROJECT environment variable.");
+    this.url = String.format(FCM_URL, projectId);
+  }
+
+  /**
+   * Gets the {@link FirebaseMessaging} instance for the default {@link FirebaseApp}.
+   *
+   * @return The {@link FirebaseMessaging} instance for the default {@link FirebaseApp}.
+   */
+  public static FirebaseMessaging getInstance() {
+    return getInstance(FirebaseApp.getInstance());
+  }
+
+  /**
+   * Gets the {@link FirebaseMessaging} instance for the specified {@link FirebaseApp}.
+   *
+   * @return The {@link FirebaseMessaging} instance for the specified {@link FirebaseApp}.
+   */
+  public static synchronized FirebaseMessaging getInstance(FirebaseApp app) {
+    FirebaseMessagingService service = ImplFirebaseTrampolines.getService(app, SERVICE_ID,
+        FirebaseMessagingService.class);
+    if (service == null) {
+      service = ImplFirebaseTrampolines.addService(app, new FirebaseMessagingService(app));
+    }
+    return service.getInstance();
+  }
+
+  /**
+   * Sends the given {@link Message} via Firebase Cloud Messaging.
+   *
+   * @param message A non-null {@link Message} to be sent.
+   * @return An {@code ApiFuture} that will complete with a message ID string when the message
+   *     has been sent.
+   */
+  public ApiFuture<String> sendAsync(@NonNull Message message) {
+    return sendAsync(message, false);
+  }
+
+  /**
+   * Sends the given {@link Message} via Firebase Cloud Messaging. If the {@code dryRun} option
```

Comment:
Insert new paragraph after the first sentence.

Reply:
Done